### PR TITLE
replaced 'user' in cmd.* invocations to 'runas'

### PR DESCRIPTION
**Pull request by lsh-0**

### CODE REVIEW

**Comment by lsh-0** [comment]
looks like some environment variables were missing and 

    /dev/tcp/${PGHOST}/${PGPORT}

became

    /dev/tcp//

**Comment by lsh-0** [comment]
it's all this stuff that's failing: https://github.com/elifesciences/elife-xpub-formula/blob/master/salt/elife-xpub/init.sls#L34-L76

because Salt doesn't execute states from a login shell. When it switches users nothing in `/etc/profile.d/` is going to get sourced - see [this](http://www.linuxfromscratch.org/blfs/view/6.3/postlfs/profile.html#etc-profile-profile) and [this](https://github.com/saltstack/salt/issues/45556).

Not sure how you managed to get it to work before. I'm going to have to get really familiar with this formula if you want me to fix it.

**Comment by lsh-0** [comment]
> if you want me to fix it.

I'm crabby. I don't like arcane bash mysteries.

**Comment by giorgiosironi** [comment]
Lack of `/etc/profile` inclusion is a classic problem of Unix automation. Will take a look.

**Comment by giorgiosironi** [comment]
This has wider implications though, such as `ENVIRONMENT_NAME` disappearing:
https://github.com/elifesciences/builder-base-formula/blob/5d8102bed07710a7bc3a41d8c57576c789c0c0c7/elife/environment-name.sls

**Comment by giorgiosironi** [comment]
Starting from reproducing this in Vagrant. Suspect `sudo` won't cut it as it would have the environment variables like in https://github.com/elifesciences/issues/issues/5438

**Comment by giorgiosironi** [comment]
It does reproduce:
```
elife@dev--elife-xpub:~$ sudo salt-call state.highstate
...
              stdout:
                  699a2c5cd0d8b67c25c426a2603426774819998a: Pulling from elifesciences/elife-xpub
                  Digest: sha256:77ac322ae3a673ed2ad43adcaecb8e81174fe31714e513a4f7107bfd4448ee08
                  Status: Downloaded newer image for elifesciences/elife-xpub:699a2c5cd0d8b67c25c426a2603426774819998a
                  bash: : Name or service not known
                  bash: /dev/tcp//: Invalid argument
```

File: salt/elife-xpub/init.sls [hunk 1]
```diff
@@ -102,10 +102,8 @@ elife-xpub-database-available:
 
     cmd.run:
         - name: /usr/local/bin/wait-database.sh
-        - user: {{ pillar.elife.deploy_user.username }}
-        - cwd: /srv/elife-xpub
-        - user: {{ pillar.elife.deploy_user.username }}
         - cwd: /srv/elife-xpub
+        #- runas: {{ pillar.elife.deploy_user.username }}
```
Comment (giorgiosironi) [inline salt/elife-xpub/init.sls]: when uncommenting this the environment variables appear, at least with my `sudo` command, but then I start seeing this in the `postgres` container logs:
```
postgres_1     | Error: Database is uninitialized and superuser password is not specified.
postgres_1     |        You must specify POSTGRES_PASSWORD to a non-empty value for the
postgres_1     |        superuser. For example, "-e POSTGRES_PASSWORD=password" on "docker run".
postgres_1     | 
postgres_1     |        You may also use "POSTGRES_HOST_AUTH_METHOD=trust" to allow all
postgres_1     |        connections without a password. This is *not* recommended.
postgres_1     | 
postgres_1     |        See PostgreSQL documentation about "trust":
postgres_1     |        https://www.postgresql.org/docs/current/auth-trust.html
```

Comment (giorgiosironi) [reply salt/elife-xpub/init.sls]: https://github.com/docker-library/postgres/issues/681

Comment (giorgiosironi) [reply salt/elife-xpub/init.sls]: Vagrant testing seems to show that https://github.com/elifesciences/elife-xpub-deployment/pull/46 solves the problem even when running from a root shell.

Comment (giorgiosironi) [reply salt/elife-xpub/init.sls]: Vagrant testing seems to show that https://github.com/elifesciences/elife-xpub-deployment/pull/46 solves the problem even when running from a root shell.

Comment (giorgiosironi) [reply salt/elife-xpub/init.sls]: Vagrant testing seems to show that https://github.com/elifesciences/elife-xpub-deployment/pull/46 solves the problem even when running from a root shell.

**Comment by giorgiosironi** [comment]
Let this here as I get the fixes reviewed by @elifesciences/xpub people.

**Comment by giorgiosironi** [comment]
Let this here as I get the fixes reviewed by @elifesciences/xpub people.

**Comment by giorgiosironi** [comment]
Latest `master` should now have the Postgres container fixes.

**Comment by giorgiosironi** [comment]
Some memory problems to reproduce in Vagrant.

**Comment by giorgiosironi** [comment]
Have to extend to 4GB because the Java applications are getting to the 2GB limit.

**Comment by giorgiosironi** [comment]
Note that in https://alfred.elifesciences.org/job/pull-requests-projects/job/elife-xpub-formula/view/change-requests/job/PR-78/2/console the variables were there:
```
05:50:00  [54.146.197.47] out:           ID: elife-xpub-database-available
05:50:00  [54.146.197.47] out:     Function: cmd.run
05:50:00  [54.146.197.47] out:         Name: /usr/local/bin/wait-database.sh
05:50:00  [54.146.197.47] out:       Result: False
05:50:00  [54.146.197.47] out:      Comment: Command "/usr/local/bin/wait-database.sh" run
05:50:00  [54.146.197.47] out:      Started: 05:49:13.967752
05:50:00  [54.146.197.47] out:     Duration: 45279.746 ms
05:50:00  [54.146.197.47] out:      Changes:   
05:50:00  [54.146.197.47] out:               ----------
05:50:00  [54.146.197.47] out:               pid:
05:50:00  [54.146.197.47] out:                   23366
05:50:00  [54.146.197.47] out:               retcode:
05:50:00  [54.146.197.47] out:                   124
05:50:00  [54.146.197.47] out:               stderr:
05:50:00  [54.146.197.47] out:                   Starting xpub_postgres_1 ... 
05:50:00  [54.146.197.47] out:                   ?[1A?[2K
05:50:00  [54.146.197.47] out:                   Starting xpub_postgres_1 ... ?[32mdone?[0m
05:50:00  [54.146.197.47] out:                   ?[1BPulling app (elifesciences/elife-xpub:699a2c5cd0d8b67c25c426a2603426774819998a)...
05:50:00  [54.146.197.47] out:               stdout:
05:50:00  [54.146.197.47] out:                   699a2c5cd0d8b67c25c426a2603426774819998a: Pulling from elifesciences/elife-xpub
05:50:00  [54.146.197.47] out:                   Digest: sha256:77ac322ae3a673ed2ad43adcaecb8e81174fe31714e513a4f7107bfd4448ee08
05:50:00  [54.146.197.47] out:                   Status: Downloaded newer image for elifesciences/elife-xpub:699a2c5cd0d8b67c25c426a2603426774819998a
05:50:00  [54.146.197.47] out:                   bash: postgres: Name or service not known
05:50:00  [54.146.197.47] out:                   bash: /dev/tcp/postgres/5432: Invalid argument
...
```

**Comment by giorgiosironi** [comment]
Now
```
    elife-xpub--vagrant: [INFO    ] Executing command '/usr/local/bin/wait-database.sh' as user 'elife' in directory '/srv/elife-xpub'
```
hangs indefinitely.

**Comment by giorgiosironi** [comment]
...because it's pulling huge Docker images.

**Comment by giorgiosironi** [comment]
https://github.com/saltstack/salt/issues/45556 describes some possible solutions by the way.

**Comment by lsh-0** [comment]
> because the Java applications are getting to the 2GB limit.

the jvm will use a slightly convoluted algorithm to determine what *percentage* of the available memory to reserve unless a fixed amount is explicitly specified: https://docs.oracle.com/javase/8/docs/technotes/guides/vm/gctuning/parallel.html#default_heap_size

above 1GB of ram the amount reserved is not a huge amount of RAM but it will grow a bit more

**Comment by lsh-0** [comment]
This is looking much healthier. Is `pr-base` salvageable or is it ready to merge as-is?

File: salt/elife-xpub/init.sls (hunk 1 above)
Comment (lsh-0) [reply salt/elife-xpub/init.sls]: that PR has been merged now

**Comment by giorgiosironi** [comment]
Sourcing the profile was a long shot, but is a stable workaround. Let's go for it.

**Comment by lsh-0** [comment]
cheers, @giorgiosironi . I'll see that elife-xpub has it's salt updated today

**Pull request merged**